### ISSUE TITLE
2.0 blockers

### ISSUE THREAD

**Issue by bcardarella** [DockYard/ember-easy-form#196]
- [ ] block form with HTMLBars
- [ ] use [scoped helpers](https://github.com/emberjs/ember.js/pull/10244)
- [ ] ember-cli-ify
- [ ] track success state
- [ ] `input` should be useable in isolation outside of `form-for` blocks

**Comment by bcardarella** [comment]
I'm aiming to start this in February.

**Comment by joostdevries** [comment]
:+1: Just wondering, what's the reason you've chosen to use helpers instead of components for this? I'm trying to figure out where to draw that line myself :-)

**Comment by bcardarella** [comment]
@joostdevries the helper will likely delegate to a component. But I don't want to overwrite Ember's own `input` helper yet I want to maintain the same API with EasyForm's `input` helper.

**Comment by Samsinite** [comment]
Why not depreciate use of {{input}} as a overridden easyForm helper? Then it gives people an easy upgrade path for 2.0 by upgrading to the depreciated version, fixing the warnings, then upgrading to 2.0

**Comment by Asgaroth** [comment]
agree with @Samsinite overriding the ember `{{input}}` is causing trouble, and is preventing me to update my app to ember@1.10.0

Having a separate helper could avoid any conflicts.

**Comment by Samsinite** [comment]
@bcardarella Is there any reason why you can't bump ember-easy-form up a minor revision, create say an `{{ef-input}}` component and deprecate easy-forms  `{{input}}` helper? Ember's `{{input}}` helper should be useable both inside of and outside of `{{form-for}}` blocks. Otherwise `{{form-for}}` is completely unusable with other component input helpers that might user `{{input}}`. Overriding any ember helper is a very bad idea, and forced me to ditch using <strong>ember-easy-form</strong> in projects months ago.

**Comment by bcardarella** [comment]
yes, I don't believe that is the proper way forward. As indicated in the blockers listed above the scoped helpers are the correct path forward.

**Comment by Asgaroth** [comment]
How can I help? Any pointers to get started?

**Comment by bcardarella** [comment]
@Asgaroth I think it is reasonable to suggest a work-around for the time being but I don't think that work-around should be in this library. Instead it should probably be a matter of openeing existing classes and sub-classing to get around the issue. If someone could provide me with an example repo that reproduces the issue I can probably get the work-around to everyone quickly

**Comment by Asgaroth** [comment]
@bcardarella this one reproduces http://emberjs.jsbin.com/bazofuyewu/2/edit

**Comment by piotrekbator** [comment]
@bcardarella are you working on easy-form 2.0? Can I see it's code anywhere? Can I help?

The current version seems not to be compatible with htmlbars and ember 1.10 and it's blocking the otherwise smooth upgrade path. I'd love to help speed things up.

**Comment by bcardarella** [comment]
@piotrekbator the primary blocker is waiting on scoped helpers. That being said if someone could explore a 1.x patch for ember 1.10 + htmlbars that would be quickly merged and released

**Comment by piotrekbator** [comment]
@bcardarella thanks for answering. I'll try then to have a look at this on the weekend.

**Comment by mikeumus** [comment]
:+1:

**Comment by g-cassie** [comment]
@piotrekbator Are you working on a patch? Let me know if not and I can try to take a look.

**Comment by piotrekbator** [comment]
@g-cassie well I gave it a try, but it looks like it'd require quite a rewrite rather than a patch. HTMLBars has a new API for helpers which is especially problematic when overwriting the input helper lik easyform does. Barber doesn't cooperate too, and debugging it (unsuccessfully) took me a lot of time - it'd be probably easier to just dump it and convert the library to an ember-addonish thing. But this idea seems to be unpopular judging from discussions in other issues...

So for now I think I'll stick with ember 1.9 and handlebars (if I'll be able to upgrade from 1.8 - I'm having easyform based problems with this too).

But maybe moving easyform development from ruby to JS platform before 2.0 is not such a bad idea. We could move it to newer tools without changing the library itself - allow easy installation in ember-cli and make development easier and then use this plaftorm to upgrade to 2.0. What do you think @bcardarella?

**Comment by jcope2013** [comment]
@piotrekbator what problems you having upgrade from ember 1.8 to ember 1.9? this build seemed to do it https://gist.github.com/jwahdatehagh/f06fab8037965f788b0c, I was able to upgrade with easyform so let me know if you have any questions

**Comment by g-cassie** [comment]
@plotrekbator Just came across the below comment in another issue in this repo.  I'm going to give it a shot to see if I can get my app working on 1.10.

https://github.com/dockyard/ember-easy-form/issues/175#issuecomment-74383722

**Comment by ulisesrmzroche** [comment]
Is anyone working on this right now? I got free time to help.

**Comment by jcope2013** [comment]
@ulisesrmzroche 

if you want to use on 1.10, add these templates under an easy-form folder in your templates directory

https://github.com/sir-dunxalot/ember-easy-form-extensions/tree/master/app/templates/easy-form

and use this file in your vendor folder or wherever, I renamed the easy form input to ef-input to not conflict with the ember input helper as I was having conflicting problems but someone could look into a solution to fix it if they liked 

https://gist.github.com/jcope2013/8f969b6e7c8244119ef8

I made some moderations to this (https://github.com/sir-dunxalot/ember-easy-form-extensions/blob/master/vendor/ember-easy-form.js#L138) as it was missing some of the latest work done on easy-form master to allow form-for to work properly, etc.

**Comment by ulisesrmzroche** [comment]
Yeah, let's just fix this permanently.  Especially the input helper. Are you going to be working on it still?

**Comment by ulisesrmzroche** [comment]
I think the easy form input helper is a custom helper right? Not the native ember one.

**Comment by jcope2013** [comment]
@ulisesrmzroche yeah i have also have an input-unbound easyform helper that i am going to be looking into fixing, prolly tomorrow and i can look around more at the input helper also, and yeah the easy form input helper is a wrapper around the ember input helper with additional functionality, in previous versions, you can use  the regular ember input  outside form-for and the easy form one inside form-for but that broke in 1.10 so more fixes need to be made on top of what has already been done

**Comment by ulisesrmzroche** [comment]
Yeah, I've worked on this before, kind of a long time ago though. Let me
dig around and see what's up.
ᐧ

**Comment by jme783** [comment]
@jcope2013 Thanks so much for your work on these shims, I just integrated them into my project and it's working out well for me. Would love to stay in the loop as to if there are any official updates to ember-easy-form over the next few weeks for supporting Ember 1.10

**Comment by tspacek** [comment]
@jcope2013 Thanks for your shim, it has got me the furthest towards getting ember-easy-form to work in 1.10 so far. Does the input ef-helper pick up the type of its property correctly for everyone else? It doesn't show a checkbox for bools for example for me, perhaps due to ED beta 15.

**Comment by jcope2013** [comment]
@tspacek ok so i was having problems with the "as" argument also, i will have to investigate more later, that might not be functioning properly in the current state or it just might be checkboxes that are broken, since there is there if statement related to checkbox in the input method https://github.com/emberjs/ember.js/blob/v1.10.0/packages/ember-handlebars/lib/controls.js#L209

**Comment by fsmanuel** [comment]
@jcope2013 great work around!

**Comment by billybonks** [comment]
@jcope2013 this only works if you pass your contextual element via paramters, any idea on how to get contextual variables to evaluate properly

{{ef-input myValue placeholder=myPlaceholder}} the plave holder will be the string myPlaceholder}}

**Comment by billybonks** [comment]
so i have gotten pretty close to solving the problem
```javascript
Ember.HTMLBars._registerHelper('input', Ember.HTMLBars.makeBoundHelper(function(property, hash,options,env) {

  return Ember.Handlebars.helpers.view.helperFunction.call(this,[Ember.EasyForm.Input], hash, options,env);
  //return EasyFormShims.viewHelper(this, Ember.EasyForm.Input, options);
}));
```
![screen shot 2015-03-24 at 7 36 49 pm](https://cloud.githubusercontent.com/assets/1640136/6800828/373bd7a4-d25d-11e4-8dc6-a8d6bab35f36.png)

im calling the helper like this
`{{input wow placeholder=wow}}`

is it possible to render the view from here? hash contains the bound value property contains the

**Comment by g-cassie** [comment]
I'm not sure if you are describing the same issue I had, but taking a look at this code helped me.

```
  var modelPropertyPath = function(property) {
    if(!property) { return null; }

    var startsWithKeyword = !!options.data.view._keywords[property.split('.')[0]];;

    if (startsWithKeyword) {
      return property;
    }

    if (modelPath) {
      return modelPath + '.' + property;
    } else {
      return property;
    }
  };
```

Basically if you want to access an attribute that is not on the model then you need to start the reference with a keyword - for example `placeholder=controller.placeholder`.  I think the inspector showing the properties as a string is a red herring.  If you look into the innards you can see that the properties get bound when the view is instantiated.

**Comment by billybonks** [comment]
this jsbin recreates the problem i am having 

http://emberjs.jsbin.com/bubiwe/5/edit

**Comment by g-cassie** [comment]
@billybonks I'll take a more detailed look tomorrow. I was binding to `selection` which I realize now behaves differently. Try adding this to easyForm:
```
    options.hash.placeholderBinding     = modelPropertyPath(options.hash.placeholder);
```
Immediately after these lines:
```
options.hash.contentBinding   = modelPropertyPath(options.hash.collection); options.hash.selectionBinding = modelPropertyPath(options.hash.selection);
options.hash.valueBinding     = modelPropertyPath(options.hash.value);
```
Haven't tested this.

**Comment by billybonks** [comment]
tested it doesn't make a difference
`the value binding isn't working even though its line exists`
ill also be taking a more detailed look tomorrow, given these hints

**Comment by g-cassie** [comment]
@billybonks I got your jsbin to work. Trick was to add "Binding" after each property name.

http://jsbin.com/zatupijivo/1/

**Comment by billybonks** [comment]
@g-cassie  thanks

**Comment by brian-gates** [comment]
Any plans to officially support 1.10, 1.11, or 1.12 anytime soon?

**Comment by bcardarella** [comment]
@brian-gates there is no plan to support those versions

**Comment by brian-gates** [comment]
Are you just ignoring 1.x until 2.0 comes along?

**Comment by rwjblue** [comment]
@bcardarella and I have discussed it a little bit, and I think we could support roughly (we can't override `{{input}}` though) the current API for 1.10+ but I think he wants to wait for https://github.com/emberjs/rfcs/pull/64 before rewriting.

**Comment by brian-gates** [comment]
That's fine by me; I'm already using `{{ef-input}}` as per one of the suggested hacks above.  This lib is the only one holding me up from upgrading past 1.10 atm.

**Comment by bcardarella** [comment]
@brian-gates the upgrade path for addons is such a pain between Ember 1.10 - 1.13 that I simply don't have the time to roll custom versions along the way. If someone wanted to take a stab at it so we can have a bridge version I'm happy to merge in those changes but its not something that I can personally dedicate time to. I'd rather focus on the 2.x series as that has some guarantee of an easier life for addon devs

**Comment by bcardarella** [comment]
btw, the requirement for this bridge version is to not break semver with the current 1.x series of this addon. That means the accepted PR would have to work with HTMLBars *and* override `input`. It should be a drop-in version and wouldn't require any changes to the apps consuming the library.

**Comment by brian-gates** [comment]
Isn't ember 2.0 still quite a long way away?  Or are you talking about 2.0 of this lib?

**Comment by rwjblue** [comment]
@brian-gates - Ember 2.0.0 should ship on 2015-07-24, but it will likely not be supported by this lib either (unless we want to use an interim API until https://github.com/emberjs/rfcs/pull/64 lands).

**Comment by bcardarella** [comment]
Once this RFC lands I'm going to immediately start the work. I don't believe it will take me more than a week to get the full rewrite done. This is also affecting our client apps as well as some personal projects I have. I feel the pain.

**Comment by brian-gates** [comment]
Ah, 1 month certainly seems reasonable.  

One concern I have is that jumping directly from Ember 1.9 to 2.0 seems like it will be more painful without all of the nice deprecation warnings provided in 1.10+.  Do you have any plans for your app to handle that?  Or just RTFM and deal with it?

**Comment by gonvaled** [comment]
What is the status of this? Project seems abandoned? I am considering using this for my Ember 2.2 app, but not sure if I will be able to integrate it successfully.

**Comment by diogomafra** [comment]
Hi @gonvaled , we released a new version yesterday. :smile: 

The most recent version of the library is on the branch `1-0-stable`. It is compatible with Ember 1.13 and 2.\*.

You can get it on the branch `1-0-stable` or as an ember-addon [here](https://www.npmjs.com/package/ember-easy-form).

**Comment by gonvaled** [comment]
@diogomafra So sorry! I only checked the master branch, and did not realize development was ongoing on other branches. Great to see this project alive and healthy!

**Comment by diogomafra** [comment]
@bcardarella what are your goals with the 2.0 rewrite? The checklist in this PR is still valid?

I think we can support these features in the `1-0-stable` branch, and rewrite it only when Glimmer 2 lands.

- 1.0 - support the old api and scoped helpers
- 1.* - deprecate the old api
- 2.0 - remove the old api and rewrite the components as [Glimmer Components](https://github.com/emberjs/ember.js/issues/12129)

Rewriting it now and again in a few months does not seem like a good idea. :smile:

**Comment by bcardarella** [comment]
@diogomafra somewhat valid. I want to lean more heavily on ember-async-button and ember-one-way-controls if necessary.

Basically, I'd like to have ember-easy-form be a best-in-class form builder and to do that I think we should leverage some already created solutions for some of the control level stuff

**Comment by bcardarella** [comment]
Closing in favor of #229
